### ISSUE TITLE
remove FileContentsHolder module as FileContents object is available for filters on TreeWalker in TreeWalkerAudit Event 

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3573]
This class - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/FileContentsHolder.java

The only reason of FileContentsHolder is to keep FileContents.
This is a HACK to keep file content as global variable to be accessible from all other places (class that wrap one field and provide simple getter for it).

**Migration note for checkstyle users:**
So FileContentsHolder is not required any more and just need to be removed from clients configurations.
In scope of  #4714 , we moved FileContents object to [TreeWalkerAuditEvent](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java#L34).

**Comment by rnveach** [comment]
> is HACK 

My exact wording. :)
My proposal inside [MultiThreaded Checkstyle thread](https://groups.google.com/forum/?hl=en#!topic/checkstyle-devel/ay2CxvaKeQo) was to merge this behavior into `FileContents` and pass that around to the Checks and Filters instead of `FileText`. We could expand on this and give `FileContents` special storage to allow Checks their own custom area.

**Comment by MEZk** [comment]
@romani 
Added to GSoC as it blocks https://github.com/checkstyle/checkstyle/issues/4382 and the solution based on the @rnveach's work will be a good starting point.

**Comment by MEZk** [comment]
What should be done to resolve the issue:

**Step 1**
1. Add the new class which will store FileText object and a set of localized messages. Proposed name is 'FileProcessingResult' (or @rnveach's class name).
2. MessageDispatcher#fireErrors should receive an object of class FileProcessingResult. The signature of the public API is required to be changed:

```java
void fireErrors(String fileName, FileProcessingResult fileProcessingResult);
```

3. Fix AbstractFileSetCheck and Checker to work with the new method (fireErrors) signature.
4. Change the signature of Checker#processFile. The method should return FileProcessingResult object.
5. Fix Checker tests.
6. Fix TranslationCheck.

**Step 2**
1. Replace FileText with FileContents in FileProcessingResult class.
2. Change signature of FileSetCheck#process in order to receive FileContents instead of List<String> lines.
3. Fix AbstractFileSetCheck.
4. Fix all classes which implements FileSetCheck ('process' method signature).
5. Fix failing tests.
6. Add FileContents to audit event to make it available for filters.

**Step 3**
1. Remove FileContentsHolder (do not forget to remove it from configs, documentation, tests).
2. Fix SuppressionCommentFilter. Remove 'fileContentsReference', 'getFileContents', 'setFileContents'. Change 'accept', 'tagSuppressions' methods.
3. Fix SuppressWithNearbyCommentFilter.
4. Fix failing tests.

Each step should be a separate PR.

@timurt 
Please, start working on this task. You can use @rnveach's [branch](https://github.com/rnveach/checkstyle/commits/issue_138) as a starting point.

**Comment by rnveach** [comment]
There is also alot of steps for just 1 issue, which makes think we should break this up somehow if we do decide to do all this.
I'm not sure we can review all this in 1 PR too.

> 'FileProcessingResult'

I feel fixing current design of `FileContentsHolder` shouldn't include adding `FileProcessingResult` which is a big change by itself.
Please also allow @romani some time to review design decisions as we sometimes disagree.

**Comment by MEZk** [comment]
@rnveach @timurt  
Ok, lets wait for @romani .

**Comment by romani** [comment]
@rnveach , @MEZk , @timurt , sorry for delay.

I did review of Steps 1-3, they looks ok, I do not have time to make deep analysis ...
I am ok to keep all in oner issue, but there should be at least 3 PRs (any number of PRs to ease code-review of changes.)

I will do 3.8.3 hot fix release till the end of week, and do upgrade to checkstyle 8.0 , so master branch will be ready for major changes in processing and APIs.

**Comment by timurt** [comment]
Finished first step,
Had problem with `Checker` class, 

> Class Fan-Out Difficulty is 26

 - as I understood it is because of new class I ve created

**Comment by MEZk** [comment]
> Had problem with Checker class,

Answered in PR #4468.

**Comment by romani** [comment]
Sorry for delay...

> Change signature of FileSetCheck#process in order to receive FileContents instead of List lines.

here is conceptual problem. [FileContents](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java) is TreeWalker's class, it contains comments (stuff from java AST) it should be used only in scope of TreeWalker.
FileSetCheck should stay on list of Lines and do processing as there is no java related Checks/modules at all around.

FileContetsHolder is used only in Filters. In current implementation all filters are in Checker (list of lines  and AST modules owner), that is why such global object was used to deliver context of AST to Filters.

Checker should not know about AST at all, it operates by files and list of lines, that is all.
We need to move all AST related filters to TreeWalker, TreeWalker should become a parent for SuppressionCommentFilter and SuppressWithNearbyCommentFilter (the only filters that need FileContetsHolder).

There should be two types of filters: 
`Filter {boolean accept(AuditEvent event);}` (at it is now , for any plain file processing as plain text)
`TreeWalkerFilter {boolean accept(TreeWalkerAuditEvent event);}` (NEW)

```
TreeWalkerAuditEvent { 
  String fileName, 
  FileContext filecontext;
  DetailAst root;
  AstLocalizedMessages messages;
}
```

All events that pass filters of TreeWalker are going to filters of parent (Checker), it is better to remove any AST related context from Messages as Checker should not know/see them. We probably should make interface for LocalizedMessage and  AstLocalizedMessage extends LocalizedMessage. So Checks creates AstLocalizedMessage but to Checker we give such instances by LocalizedMessage type, so Checker(with its filters) will not see AST fields.

@MEZk , please find me in skype to discuss this.

**Comment by rnveach** [comment]
> Filter 
> TreeWalkerFilter 

Should we name everything after main modules? CheckerFilter and TreeWalkerFilter. CheckerAuditEvent and TreeWalkerAuditEvent.
Should we assign some special interface to these modules that say they are their own central point for all children communication? Something similar to `RootModule`.

> TreeWalkerAuditEvent 

We should convert `AuditEvent` to an interface or abstract class and have CheckerAuditEvent and TreeWalkerAuditEvent implement/extend it.

**Comment by romani** [comment]
> Should we name everything after main modules? CheckerFilter and TreeWalkerFilter. CheckerAuditEvent and TreeWalkerAuditEvent

yes it is better to rename all now, it make code clear of context (what class belong to what context)(context is plain text or AST).

> Should we assign some special interface to these modules that say they are their own central point for all children communication? 

I am not sure what you mean by this. TreeWalkerFilter interface should be enough.

> We should convert AuditEvent to an interface or abstract class and have CheckerAuditEvent and TreeWalkerAuditEvent implement/extend it.

ok, but lets hide implementations out of API package.

**Comment by rnveach** [comment]
> > Should we assign some special interface to these modules 

> I am not sure what you mean by this. 

I meant `TreeWalker`. Should we give it a special interface like `CentralCommunicationModule` or something saying it localizes all messages and filters inside it before passing them to parent. This would put them up on their own throne as they are not the same as checks or normal file sets anymore. This would probably more depend on if we need to recognize and process them separately from others.

**Comment by MEZk** [comment]
@timurt 
Steps 1, 2, and 3 are not required to do. Please see https://github.com/checkstyle/checkstyle/issues/4714 and [@romani's comment](https://github.com/checkstyle/checkstyle/issues/3573#issuecomment-314760202) for more details. In the scope of this issue we need to completely remove FileContentsHolder. This must be done only after https://github.com/checkstyle/checkstyle/issues/4714 when we remove its usage from the filters.

**Comment by timurt** [comment]
@MEZk @romani 

> In the scope of this issue we need to completely remove FileContentsHolder

should I directly contribute to [checktyle-tester](https://github.com/checkstyle/contribution/tree/master/checkstyle-tester) or apply patch inside `wercker.yml`?

**Comment by romani** [comment]
fix is merged.
